exonum/src/runtime/dispatcher/mod.rs

```diff
@@ -523,21 +536,27 @@ impl Dispatcher {
         let instance = self
             .get_service(call_info.instance_id)
             .ok_or(CoreError::IncorrectInstanceId)?;
-        let context = ExecutionContext::for_transaction(self, fork, instance, tx.author(), tx_id);
+
+        let mut should_rollback = false;
+        let context = ExecutionContext::for_transaction(
+            self,
+            fork,
+            &mut should_rollback,
+            instance,
+            tx.author(),
+            tx_id,
+        );
 
         let mut res = runtime.execute(context, call_info.method_id, &tx.as_ref().arguments);
+        if should_rollback && res.is_ok() {
+            res = Err(CoreError::IncorrectCall.into());
+        }
+
         if let Err(ref mut err) = res {
             fork.rollback();
 
-            err.set_runtime_id(runtime_id).set_call_site(|| {
-                CallSite::new(
-                    call_info.instance_id,
-                    CallType::Method {
-                        interface: String::new(),
-                        id: call_info.method_id,
-                    },
-                )
-            });
+            err.set_runtime_id(runtime_id)
+                .set_call_site(|| CallSite::from_call_info(call_info, String::new()));
```

Comment:
Nit: Can't `""` be used instead of `String::new()` here? IMO, it would be better for readability.